### DISCUSSION TITLE
Autostart für Optolink Splitter bring ich nicht zum laufen

### DISCUSSION THREAD

**Discussion by schnuller123** [philippoo66/optolink-splitter#63] [Q&A]
Hallo Leute!

zuerst mal mein erstes Problem
bin kein grosser Linux bediener , denke mal das ist das Hauptproblem

dann mal zu den orginal pfade:
/home/pi/optolink/    steht in der anleitung
aber habe herausgefunden, das ich geänderte pfade verwenden muss:
/home/ds/optolink-splitter-main/

ich habe den user "ds" und nicht "pi"
und es wird in der Datei "xxx" von eine Group definiert, was muss ich da haben ds oder pi?

wenn ich es manuell starte dann läuft alles, auch die MQTT ausgabe!!!!

dann auch noch den LOG für den Fehler der autostart funktion:

ds@raspberrypi:~/optolink-splitter-main $ journalctl -xe

░░ A start job for unit optolink-splitter.service has begun execution.
░░ The job identifier is 35340.
Dec 23 23:03:49 raspberrypi systemd[1152]: optolink-splitter.service: Failed to determine group credentials: No such >
Dec 23 23:03:49 raspberrypi systemd[1152]: optolink-splitter.service: Failed at step GROUP spawning /bin/sleep: No su>
░░ Subject: Process /bin/sleep could not be executed
░░ Defined-By: systemd
░░ The process /bin/sleep could not be executed and failed.
░░ The error number returned by this process is ERRNO.
Dec 23 23:03:49 raspberrypi systemd[1]: optolink-splitter.service: Control process exited, code=exited, status=216/GR>
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ An ExecStartPre= process belonging to unit optolink-splitter.service has exited.
░░ The process' exit code is 'exited' and its exit status is 216.
Dec 23 23:03:49 raspberrypi systemd[1]: optolink-splitter.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ The unit optolink-splitter.service has entered the 'failed' state with result 'exit-code'.
Dec 23 23:03:49 raspberrypi systemd[1]: Failed to start Optolink-Splitter Service.
░░ Subject: A start job for unit optolink-splitter.service has failed
░░ Defined-By: systemd

**Comment by philippoo66** [comment]
Moin @schnuller123 !

Schön wäre gewesen, wenn du grad das systemd Skript hier rein gestellt hättest, das du benutzt, und auch gesagt hättest, ob du mit virtueller Umgebung (venv) arbeitest, oder nicht.

wenn ich das skript aus [Wiki](https://github.com/philippoo66/optolink-splitter/wiki/510-error:-externally%E2%80%90managed%E2%80%90environment-%E2%80%90%E2%80%90-venv) (ohne venv) an deine Infos anpasse, sieht es so aus:

```
[Unit]
Description=OptolinkVS2 Switch Service
After=network-online.target

[Service]
Type=simple
ExecStart=/usr/bin/python3 /home/ds/optolink-splitter-main/optolinkvs2_switch.py
WorkingDirectory=/home/ds/optolink-splitter-main/
Restart=always
RestartSec=3
User=ds
Group=ds

[Install]
WantedBy=multi-user.target
```

Vielleicht war es das ja schon? ansonsten müsstest du mal `groups ds` eingeben und die Ausgabe posten und ggf. Angaben zur venv machen.

Grüsse!
Phil

**Comment by schnuller123** [comment]
Hallo 
toll wie schnell du darauf antwortest
ich hab ihn nochmal komplett neu aufgesetzt und auch den User pi gelassen
so schaut der script aus, hoffe du meinst den!?!

[Unit]
Description=Optolink-Splitter Service
After=network-online.target

[Service]
Type=simple
ExecStartPre=/bin/sleep 30
User=pi
Group=pi
WorkingDirectory=/home/pi/optolink-splitter-main/
Environment="PATH=/home/pi/optolink-splitter-main/myvenv/bin"
ExecStart=/home/pi/optolink-splitter.main/myvenv/bin/python /home/pi/optolink-splitter/optolinkvs2_switch.py
Restart=always
RestartSec=3

[Install]
WantedBy=multi-user.target


der Autostart geht leider nicht, nur wenn ich ihn im Putty manuell mit folgenden Befehlen starte:

Beim booten zeigt er zuerst IP 127.0.1.1 an erst nach ca 1Min dann die richtige!! = Fritzbox DHCP?
Dann mach ich 
cd optolink-splitter-main 			den passenden Pfad einstellen
source myvenv/bin/activate			Python Virtuell starten
python optolinkvs2_switch.py			Python Tool starten
und schon kommen am MQTT Explorer Daten an und Putty zeigt sie an!
nur wenn ich den Putty beende, dann stoppen die ausgaben!

nochmal Danke !!!!!!

**Reply by philippoo66** [reply]
moin!

hm, ok, wir haben also wohl noch zwei kleine issues. Das erste ist ganz einfach, da sind nur noch zwei kleine Tippfehler:

![grafik](https://github.com/user-attachments/assets/e06d3268-732a-40ca-8b37-74f35ab82db7)

es muss heissen:
`ExecStart=/home/pi/optolink-splitter-main/myvenv/bin/python /home/pi/optolink-splitter-main/optolinkvs2_switch.py`
Computer sind bei sowas immer kleinlich ;-)

Die zweite Sache scheint mit Netzwerkverbindung zu tun zu haben? Eigentich legen wir ja mit der Zeile `After=network-online.target` fest, dass der Dienst überhaupt erst starten soll, _nachdem_ das Netzwerk verbunden ist. Das sollte also eigentlich keine Problem bereiten. 

Wo zeigt es 127.0.1.1 an? Das ist eine seltsame Adresse. 127.0.0.1 wäre local host, und ohne Netzwerkverbindung hat es eigentlich erstmal eine 192er Adresse mit komischen Subnet und so.

Oben hatte es ja ausgegeben `Subject: Process /bin/sleep could not be executed`, deswegen hatte ich das weggelassen, das ist eigentlich nur nötig, wenn network-online.target schon gemacht ist, aber noch auf eine WLAN Verbindung gewartet werden soll...

Bitte korrigier mal die ExecStart Zeile und probier es damit noch mal. Du kannst ja den Dienst auch 'manuell' starten (ohne Raspi re/boot) mit `sudo systemctl start optolink-splitter`. Wenn es dabei Fehler gibt, kopiert den Output bitte nochmal hier rein!

ps: 
> nur wenn ich den Putty beende, dann stoppen die ausgaben!

das ist normal, weil Putty ist in dem Fall der 'Host Prozess', und wenn du den beendest, wird auch das Programm beendet.

**Comment by schnuller123** [comment]
Hallo
wahnsinn dein Support!!!!

127.0.1.1 kommt am Monitor der direkt am Pi Zero hängt
bediene den Zero aber nur über Putty.

nach ca 1Min blitz er kurz auf ( Clear ? ) und zeigt dann 192.168.178.164 an

Eintrag habe ich geändert, aber er startet immer noch nicht

sudo systemctl daemon-reload
sudo systemctl start optolink-splitter.service
ohne Meldung

sudo systemctl status optolink-splitter.service 
sagt wieder:

pi@raspberrypi:~ $ sudo systemctl status optolink-splitter.service
● optolink-splitter.service - Optolink-Splitter Service
     Loaded: loaded (/etc/systemd/system/optolink-splitter.service; enabled; pr>
     Active: activating (start-pre) since Wed 2024-12-25 13:25:20 CET; 2s ago
  Cntrl PID: 679 (sleep)
      Tasks: 1 (limit: 387)
        CPU: 20ms
     CGroup: /system.slice/optolink-splitter.service
             └─679 /bin/sleep 30

dann wieder 
cd optolink-splitter-main 			den passenden Pfad einstellen
source myvenv/bin/activate			Python Virtuell starten
python optolinkvs2_switch.py			Python Tool starten

eingegeben und MQTT kommen Daten

**Reply by philippoo66** [reply]
der output den du hinkopiert hast sagt aber dass der Splitter (naja, erstmal der Dienst) läuft:
```
Cntrl PID: 679 (sleep)
Tasks: 1 (limit: 387)
CPU: 20ms
```
du muss halt erstmal ne halbe Minute abwarten wenn du das sleep 30 noch drin hast. 
```
CGroup: /system.slice/optolink-splitter.service
└─679 /bin/sleep 30
```
ne halbe Minute kann lang werden wenn man wartet... ;-) und auch wenn das Skript anläuft braucht es noch ein paar Sekunden bevor die ersten Werte auf MQTT kommen

**Reply by FransOv** [reply]
Die service startet nicht in virtual Environment. 
ExecStart=/home/pi/optolink-splitter.main/myvenv/bin/python /home/pi/optolink-splitter/optolinkvs2_switch.py startet die venv nicht, kein Source Kommando. So wenn paho.mqtt nicht installiert ist aussen den venv (mit sudo) wird das nicht klappen. Auch werde ich ein WorkingDirectory in die service Definition aufnehmen.

**Reply by philippoo66** [reply]
bist du sicher?? ich starte alle meine Skripte im venv so und das funktioniert definitv. wenn es nicht im venv liefe, würden jede menge libs fehlen...

**Reply by philippoo66** [reply]
mit
```
Environment="PATH=/home/pi/optolink-splitter-main/myvenv/bin"
ExecStart=/home/pi/optolink-splitter-main/myvenv/bin/python 
```
bist du im venv. da braucht es kein extra `source ...` um noch irgendwas zu starten

**Reply by FransOv** [reply]
https://docs.python.org/3/library/venv.html

**Comment by schnuller123** [comment]
mir ist das alles zu hoch....
warum muss man eigentlich ein einer virtuellen umgebung Python starten???

kann ich das nicht "normal" starten?

ich muss im Putty nur folgende sachen starten, dann geht es:

cd optolink-splitter
python -m venv myvenv 
source myvenv/bin/activate
python optolinkvs2_switch.py

mfg Sigi

**Comment by schnuller123** [comment]
habe übrigens nur nen Zero mit folgender aktuellen Software drauf

PRETTY_NAME="Raspbian GNU/Linux 12 (bookworm)"
NAME="Raspbian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"

**Comment by FransOv** [comment]
Versuch mal im putty:
cd optolink-splitter
python optolinkvs2_switch.py

Wenn er sagt das es kein pyserial und/oder paho-mqtt gibt:
sudo pip install pyserial
sudo pip installl paho-mqtt

Danach soll es ohne venv auch marschieren.  
Die entsprechenden Zeilen im Service Definition können dan auch geändert werden..

**Comment by schnuller123** [comment]
Hallo FransOv

habs gemacht aber:

pi@raspberrypi:~/optolink-splitter-main $ 	python optolinkvs2_switch.py
Traceback (most recent call last):
File "/home/pi/optolink-splitter-main/optolinkvs2_switch.py", 
line 19, in <module>    
import serial
ModuleNotFoundError: No module named 'serial'

in der Scriptdatei optolinkvs2_switch.py
steht an Zeile 19:	import Serial

wenn ich dann sudo pip install pyserial eingebe
kommt: 	sudo: pip: command not found

**Reply by FransOv** [reply]
Versuche mal:
sudo python -m pip install pyserial
sudo python -m pip install paho-mqtt

**Comment by schnuller123** [comment]
Nein , ohne Erfolg

pi@raspberrypi:~ $ cd optolink-splitter-main
pi@raspberrypi:~/optolink-splitter-main $ sudo python -m pip install pyserial
/usr/bin/python: No module named pip

pi@raspberrypi:~/optolink-splitter-main $ cd ..
pi@raspberrypi:~ $ sudo python -m pip install pyserial
/usr/bin/python: No module named pip

auch innerhalb der virtuellen Umgebung geht es nicht

**Reply by FransOv** [reply]
Deine Version von Python hat keine Pip installiert. 
Vielleicht werd 
sudo python -m ensurepip 
helfen.

**Reply by schnuller123** [reply]
ich dreh mich im kreis:

pi@raspberrypi:~/optolink-splitter-main $ sudo python -m ensurepip
ensurepip is disabled in Debian/Ubuntu for the system python.

Python modules for the system python are usually handled by dpkg and apt-get.

    apt install python3-<module name>

Install the python3-pip package to use pip itself.  Using pip together
with the system python might have unexpected results for any system installed
module, so use it on your own risk, or make sure to only use it in virtual
environments.

ich denke es muss python zuerst mal komplett installiert sein, habe es schon versucht
bin aber gescheitert! oder man verwendet "venv"

**Reply by philippoo66** [reply]
also eins nach dem anderen...

wahrscheinlich wäre der aufruf von `pip3` hier die Lösung, `pip` ist nur ein Softwarelink auf `pip3` und der fehlt wahrscheinlich.

Aber das ist ja egal, weil pyserial und paho sind ja schon installiert, sonst würde der Splitter bei manuellem Start nicht laufen.

Also... Der Splitter läuft bei mir auf einem älteren System wo pyseriel und paho 'global' installiert sind, deswegen kann ich es hieran nicht demonstrieren, aber ich habe ein anderes Skript, das ebenfalls serial und das gleich mqtt_util Modul benutzt. 

Starte ich das Skript einfach so, erhalte ich erwartungsgemäss

![grafik](https://github.com/user-attachments/assets/e5ab0361-7d48-4ef6-ab9a-c201decb5d83)

Starte ich es mit virtueller Umgebung läuft das Skript erwartungsgemäss:

![grafik](https://github.com/user-attachments/assets/d1c1b3a7-e55b-47f6-a1e5-4c6ebc9bb6e7)

Starte ich den Dienst kommt

![grafik](https://github.com/user-attachments/assets/81db4a99-c31d-4c76-85b4-17aa57e6d9ae)


Das Dienst Skript ist folgendes:

![grafik](https://github.com/user-attachments/assets/41546ea7-a161-42b5-8c41-a9f9a244a9f2)

also analog genau das, was auch oben steht

**Comment by schnuller123** [comment]
OK jetzt habe ich mal mit
sudo apt install python3-pip
pip installiert, ist ja wohl eine anwendung die ich noch nicht hatte
dann noch 
sudo pip install pyserial
dann kommt aber:

pi@raspberrypi:~ $ sudo pip install pyserial
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    For more information visit http://rptl.io/venv

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

dann wären wir wieder bei dieser scheiss venv

hab noch versucht:
sudo apt install python3-pyserial
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package python3-pyserial

bin am ENDE!

**Reply by philippoo66** [reply]
warte mal! ein nach dem anderen! ich bin gleich bei dir! :-)

**Comment by FransOv** [comment]
Am Ende ist immer wieder eine Neustart.
Du kannst die --break-system-packages Option benutzen wie es hier erklärt wird..
https://pimylifeup.com/python-externally-managed-environment/

**Reply by philippoo66** [reply]
@schnuller123 
also lass uns festhalten: in der virtuellen Umgebung läuft das Skript, richtig?
dann brauchst du nicht installieren, weil es ist ja schon installiert.

**Reply by schnuller123** [reply]
ja klar du hast recht
aber ich bring den systemd ( autostart ) nicht hin!!!

**Reply by philippoo66** [reply]
bitte wechsel mal in dein verzeichnis `/home/pi/optolink-splitter-main/` und mach ein `ls`

**Reply by philippoo66** [reply]
da kommt jetzt wahrscheinlich unter anderem ein verzeichnis myvenv richtig?

**Reply by schnuller123** [reply]
pi@raspberrypi:~/optolink-splitter-main $ ls
 allupdate.sh
 LICENSE
 list_ports.py
 mqtt_util.py
 myvenv
 onewire_util.py
 optolink-splitter-main
 optolinkvs2.py
 optolinkvs2_switch.py
'pberrypi systemd[1]: optolink-splitter.service: Failed with result timeout.'
 __pycache__
 Python-3.8.4
 Python-3.8.4.tar.xz
 README.md
 requests_util.py
 serlog.py
 settings_ini.py
"s' exit code is 'killed' and its exit status is 15."
 tcpip_util.py
 utils.py
 viconn_util.py
 viessdata_util.py

da liegen noch tar von einer versuchten python3 install drin

**Reply by philippoo66** [reply]
ok, das ist gut. jetzt bitte
source myvenv/bin/activate

und dann
pip list

**Reply by schnuller123** [reply]
seh es selber, hab nochmal den gleichen ordner von dir drin
macht aber nichts oder?

pi@raspberrypi:~/optolink-splitter-main/optolink-splitter-main $ ls
```
allupdate.sh   mqtt_util.py     optolinkvs2_switch.py  serlog.py        utils.py
LICENSE        onewire_util.py  README.md              settings_ini.py  viconn_util.py
list_ports.py  optolinkvs2.py   requests_util.py       tcpip_util.py    viessdata_util.py
```

**Reply by schnuller123** [reply]
ja den ordner hab ich da drinls

pi@raspberrypi:~/optolink-splitter-main/myvenv $ ls
bin  include  lib  pyvenv.cfg

**Reply by philippoo66** [reply]
ah ok. also erstmal aufräumen ;-)

vielleicht löschst du den kompletten Ordner ~/optolink-splitter-main und wir fangen ganz von vorne an, ok?

**Reply by philippoo66** [reply]
am besten vor dem löschen deine settings_ini irgendwo sichern!

**Reply by schnuller123** [reply]
(myvenv) pi@raspberrypi:~/optolink-splitter-main $ pip list
Package    Version
---------- -------
paho-mqtt  2.1.0
pip        23.0.1
pyserial   3.5
setuptools 66.1.1

**Reply by philippoo66** [reply]
gut. das passt. also was wollen wir machen? aufräumen oder 'im chaos' die Lösung bewerkstelligen?

**Reply by philippoo66** [reply]
hallo!? @schnuller123 ?

**Reply by schnuller123** [reply]
sorry du hast mir viel gesendet
bin gerade am aufräumen

**Reply by schnuller123** [reply]
pi@raspberrypi:~/optolink-splitter-main $ ls
 allupdate.sh
 LICENSE
 list_ports.py
 mqtt_util.py
 myvenv
 onewire_util.py
 optolinkvs2.py
 optolinkvs2_switch.py
'pberrypi systemd[1]: optolink-splitter.service: Failed with result timeout.'
 __pycache__
 Python-3.8.4
 README.md
 requests_util.py
 serlog.py
 settings_ini.py
"s' exit code is 'killed' and its exit status is 15."
 tcpip_util.py
 utils.py
 viconn_util.py
 viessdata_util.py

nun gut so?
oder muss der Python auch runter?

**Reply by philippoo66** [reply]
ok. sag bescheid wenn du fertig bist.

**Reply by schnuller123** [reply]
ich seh gerade 2 Dateien

 optolinkvs2.py
 optolinkvs2_switch.py

sind die beide wichtig oder das gleiche Problem
wie beim Pfad

optolink-splitter und optolink-splitter-main?

**Reply by schnuller123** [reply]
die beiden sind komplett unterschiedlich
soweit bin ich fertig!
ausser es muss der Ordner /optolink-splitter-main/Python-3.8.4 
auch noch weg?

**Reply by philippoo66** [reply]
oh, warte. wieder was durcheinander...
muss nicht, kann aber weg

**Reply by philippoo66** [reply]
diese beiden
"s' exit code is 'killed' and its exit status is 15."
und 
'pberrypi systemd[1]: optolink-splitter.service: Failed with result timeout.'

sind das dateien?

**Reply by philippoo66** [reply]
sag mal hast du Teams oder so, dann können wir grad ne online session machen

**Reply by schnuller123** [reply]
hab ich auch schon gesehen, wenn ich das wüsste!
sind glaub ich log text dateien
löschen?

**Reply by philippoo66** [reply]
ja löschen.

**Reply by philippoo66** [reply]
hast du Teams oder so, dann können wir grad ne online session machen

**Comment by schnuller123** [comment]
team nein, aber Teamviewer
oder installier Teams
no problem

**Reply by philippoo66** [reply]
ok nehmen wir teamviewer

**Reply by schnuller123** [reply]
Bitte verwenden Sie die unten stehenden Zugangsdaten, um eine Session zu starten. TeamViewer ID: 1 332 223 692 Passwort: 2gd7b8zm

**Reply by philippoo66** [reply]
fuck ich habe eine alte mitdeiner nicht kompatible version

**Reply by philippoo66** [reply]
hast du einMSAccount? Teams läuft auch im Browser

**Reply by schnuller123** [reply]
nein hab ich nicht
dann lad ich halt Teams runter

**Reply by schnuller123** [reply]
dauert halt, sorry

**Reply by philippoo66** [reply]
du musst trotzdem schnell ein Account machen zum Anmelden

**Comment by schnuller123** [comment]
wollen wir auf Whatsapp gehen?
08195/999010

**Reply by philippoo66** [reply]
ok warte

**Reply by philippoo66** [reply]
ich muss grad mein Händi finden...

**Reply by schnuller123** [reply]
lad halt einfach bei Teamviewer
das QuickSupport runter, nicht gross
mein Teams macht Probleme!!

**Reply by philippoo66** [reply]
lag im Auto und Akku runter. klingel mal durch +49 1525 3996502

**Reply by schnuller123** [reply]
https://teams.live.com/l/invite/FEAzTTjiN1K1CWOdgE

**Reply by philippoo66** [reply]
eigentlich müsste sich der Splitter auch auf einen ESP portieren lassen. Es kann zwei 'Knackpunkte' geben:

1. die Sache mit dem USB für den Optolink kopf - ich weiss nicht, wie das mit USB auf dem ESP ist?
2. die paho mqtt Lib wird es wahrscheinlich nicht 1:1 für einen ESP geben?

aber ansonsten ist das Skript eigentlich sehr genügsam ;-)

**Reply by schnuller123** [reply]
ich werde es so auf dem zero lassen
danke nochmal !

**Reply by FransOv** [reply]
Ist es  gelungen die Service zu starten?
Bei mir marschiert es ohne venv. Benutze aber keine Raspi sondern eine Orange Pi 3 mit Ubuntu. Der hat Python 3.10.12 als Standard.
```
[Unit]
Description=OptolinkVS2 Switch Service
After=network.target network-online.target
StartLimitIntervalSec=30
StartLimitBurst=2

[Service]
Type=simple
ExecStartPre=/bin/sleep 30
ExecStart=/usr/bin/python3 /home/orangepi/vitocal/optolink-splitter/optolinkvs2_switch.py
WorkingDirectory=/home/orangepi/vitocal/optolink-splitter/
Restart=none
RestartSec=3
User=orangepi
# fuer Augabe auf dem Bildschirm
#StandardOutput=tty
#StandardError=tty

[Install]
WantedBy=multi-user.target
```

**Reply by philippoo66** [reply]
hi Frans!
ja, es funktioniert inzwischen alles. das `optolink-splitter.main` stand noch drin (Punkt statt Strich) und auch das `ExecStartPre=/bin/sleep 30` schien Probleme zu bereiten. Nachdem wir den Punkt korrigiert und das sleep rausgeworfen hatten, ging es :-) 
allerbesten Dank wieder für deine Unterstützung!
